### PR TITLE
pull common SQL adapters and helpers into reusable functions

### PR DESCRIPTION
**Pull request by verdverm**
…functions

### CODE REVIEW

**Comment by mitjade** [comment]
@verdverm You are doing an amazing job here. I am doing something a bit differently in https://github.com/sysgears/apollo-universal-starter-kit/pull/524. That is generating sql queries with `knex`, based on `resolverInfo`. I think combining both things could resolve into something really powerful. I initially tried integrating `join-monster` into this kit, but it did not fit in well in all the places I needed it. `knex` query builder is really powerful and above all flexible enough to do what you want.

Quick intro to my approach:
In the resolver there is a forth argument `info` for example:
```
product: async (obj, args, context, info) => ...
```
With the help of `graphql-parse-fields` you can get all query fields and then I generate all the select fields and joins so that in the end you get with the help of `knexnest` exact data that you need.
 If you are interested in what I've done, you can check `selectBy` in `https://github.com/sysgears/apollo-universal-starter-kit/blob/cli-crud/src/server/sql/helpers.js`.
To take care of one-to-many relations I needed to use `graphql-resolve-batch`, so I could pass in the `info`. But for this I need to generate a resolver to connect it all together.

Branch is currently work in progress and many things still do not yet fully work as intended.

I am officially on vacation an really promised my wife I will not do any work over the holidays. If any of this interest you we can discuss it further, but I will not really be responsive, maybe twice day at most. After the holidays I will continue on this and I can already see many things you done, that can be useful  to me in what I am doing.

**Comment by verdverm** [comment]
I was definitely thinking about how the query path will probably be needed in the `withAuth` adapter, pairing with the info comes that comes from obj/sources

**Comment by verdverm** [comment]
Check these out: 

https://github.com/sysgears/apollo-universal-starter-kit/blob/96dcc866b91004ffad3d13ff1d811c665e3a8546/src/server/sql/crud.js

https://github.com/sysgears/apollo-universal-starter-kit/pull/562/commits/96dcc866b91004ffad3d13ff1d811c665e3a8546

**Comment by mitjade** [comment]
@verdverm Looks interesting! A bit hard to wrap my head around all the code. Will take a closer look, when I have some time.

**Comment by verdverm** [comment]
I added a HOC Table based on react-table that integrates with the paging crud helper in my madd sci PR

**Comment by verdverm** [comment]
I have a hard time wrapping my head around it too!

File: src/server/sql/filters.js [hunk 1]
```diff
@@ -0,0 +1,135 @@
+import { decamelize } from 'humps';
+import { has } from 'lodash';
+
+export function currentFilter(queryBuilder, filter) {
```
Comment (vlasenko) [inline src/server/sql/filters.js]: Looks like this function is specific to `user` module. Why is it here, then?

File: docs/sql-helpers.md [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+
+In `filters.js`
+uses the current code and filter argument
+
+
+### filterBuilder
+
+uses a more advanced filtering input and processing.
+
+See https://github.com/sysgears/apollo-universal-starter-kit/issues/569
```
Comment (vlasenko) [inline docs/sql-helpers.md]: Is it okay to send user to the PR in the docs?

**Comment by verdverm** [comment]
I've upgraded this system a bunch on my mad-sci branch. Lots of other helpers for other things too. They are starting to become better together. Check the latest commit(s) over there

**Comment by mairh** [comment]
Wouldn't it make sense to pull all those upgrades into smaller PR? mad-sci branch is f*****g huge and kinda impossible to do a PR or make sense out of it.

**Comment by vlasenko** [comment]
@verdverm Are you going to take into account the points I have raised in this PR to finish it and merge?

**Comment by verdverm** [comment]
I should probably cut a new PR on this one too.

**Pull request closed without merging**